Public/Src/Engine/Processes/SandboxedProcessPipExecutor.cs

```diff
@@ -517,26 +517,85 @@ private bool IsWarning(string line)
+            {
+                Contract.Requires(linePredicate != null || regex != null);
+                LinePredicate = linePredicate;
+                Regex = regex;
+            }
+
+            /// <summary>
+            /// When <see cref="LinePredicate" /> is specified: it is invoked against <paramref name="source" /> and if 
+            /// it returns true <paramref name="source" /> is returned.
+            ///
+            /// When <see cref="Regex" /> is specified: it is invoked against <paramref name="source" /> to find all
+            /// matches; the matches are joined by <paramref name="outputSeparator"/> (or <see cref="Environment.NewLine" />
+            /// if null) and returned.
+            /// </summary>
+            internal string ExtractMatches(string source, string outputSeparator = null)
+            {
+                return LinePredicate != null
+                    ? (LinePredicate(source) ? source : string.Empty)
+                    : string.Join(outputSeparator ?? Environment.NewLine, Regex.Matches(source).Cast<Match>().Select(m => m.Value));
             }
+        }
+
+        /// <summary>
+        /// If <see cref="m_errorRegex"/> is set and its options include <see cref="RegexOptions.Singleline"/> (which means that 
+        /// the whole input string---which in turn may contain multiple lines---should be treated as a single line), returns the 
+        /// regex itself (to be used later to find all the matches in the input string); otherwise, returns a line filter 
+        /// (to be used later to match individual lines from the input string).
+        /// </summary>
+        /// <remarks>
+        /// Must not be called before <see cref="TryInitializeErrorRegexAsync"/> is called.
+        /// </remarks>
+        private OutputFilter GetErrorFilter()
+        {
+            if (m_errorRegex != null && m_errorRegex.Options.HasFlag(RegexOptions.Singleline))
+            {
+                return new OutputFilter(m_errorRegex);
+            }
+            else
+            {
+                return new OutputFilter(line =>
+                {
+                    Contract.Requires(line != null, "line must not be null.");
 
-            return m_errorRegex.IsMatch(line);
+                    // in absence of regex, treating everything as error.
+                    if (m_errorRegex == null)
+                    {
+                        return true;
+                    }
+
+                    // An unusually long string causes pathologically slow Regex back-tracking.
+                    // To avoid that, only scan the first 400 characters. That's enough for
+                    // the longest possible prefix: MAX_PATH, plus a huge subcategory string, and an error location.
+                    // After the regex is done, we can append the overflow.
+                    if (line.Length > 400)
+                    {
+                        line = line.Substring(0, 400);
+                    }

```

Comment:
I wonder if we should just remove this. It makes sense for warnings where we process every single output line of every pip. On failures we're less concerned about perf